In [1]:
import numpy as np
import pandas as pd
import scipy.stats

In [2]:
data = pd.read_csv('pilgrim.csv')

In [3]:
data.head()

ID  9Profit  9Online  9Age  9Inc  9Tenure  9District  0Profit  0Online  \
0   1       21        0   NaN   NaN     6.33       1200      NaN      NaN   
1   2       -6        0   6.0   3.0    29.50       1200    -32.0      0.0   
2   3      -49        1   5.0   5.0    26.41       1100    -22.0      1.0   
3   4       -4        0   NaN   NaN     2.25       1200      NaN      NaN   
4   5      -61        0   2.0   9.0     9.91       1200     -4.0      0.0   

   9Billpay  0Billpay  
0         0       NaN  
1         0       0.0  
2         0       0.0  
3         0       NaN  
4         0       0.0

# 1. Calculate average customer profitability with 95% confidence level

In [4]:
data_dropna = data.dropna(axis = 0,how = 'any')
profit_9 = data_dropna['9Profit']
profit_0 = data_dropna['0Profit']
def mean_confidence_interval(a, confidence=0.95):
    n = len(a)
    m, se = np.mean(a), np.std(a)/np.sqrt(len(a))
    h = se * scipy.stats.t.ppf((1 + confidence) / 2., n-1)
    print('The mean is %.2f' % m)
    print('The 95%% confidence interval is [%.2f,%.2f]' % (m-h,m+h))
    return m, m-h, m+h

In [5]:
# 1999 
mean_confidence_interval(profit_9, confidence=0.95)

The mean is 129.88
The 95% confidence interval is [126.03,133.73]


(129.878954607978, 126.02897986492788, 133.72892935102809)

In [6]:
# 2000
mean_confidence_interval(profit_0, confidence=0.95)

The mean is 154.39
The 95% confidence interval is [149.04,159.74]


(154.39448845041028, 149.0449930992727, 159.74398380154787)

# 2. (a) Evaluate if online channel has a significant impact on 1999 profitability (9Profit).

In [7]:
from sklearn.linear_model import LinearRegression
data_dropna = data.dropna(axis = 0,how = 'any')
X = data_dropna[['9Online']]
y = data_dropna['9Profit']
reg = LinearRegression().fit(X, y)
reg.score(X,y)

8.358201402347554e-05

In [8]:
from statsmodels.api import OLS
OLS(y,X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                9Profit   R-squared (uncentered):                   0.024
Model:                            OLS   Adj. R-squared (uncentered):              0.024
Method:                 Least Squares   F-statistic:                              528.7
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                   1.44e-115
Time:                        23:12:12   Log-Likelihood:                     -1.5083e+05
No. Observations:               21083   AIC:                                  3.017e+05
Df Residuals:                   21082   BIC:                                  3.017e+05
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
9Online      136.6652      5.944     22.993      0.000     125.015     148.315
==============================================================================
Omnibus:                    10692.153   Durbin-Watson:                   1.729
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            73444.507
Skew:                           2.381   Prob(JB):                         0.00
Kurtosis:                      10.805   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### From the result from the OLS regression, the R-squared is 0.024,which is very low. So online channel doesn't have a significant impact on 1999 profitability 

# 2. (b) Does age help to explain if online channel has a significant impact on 1999 profitability?



In [9]:
from sklearn.linear_model import LinearRegression
data_dropna = data.dropna(axis = 0,how = 'any')
X_2 = data_dropna[['9Age','9Online']]
y_2 = data_dropna['9Profit']
reg = LinearRegression().fit(X_2, y_2)
reg.score(X_2,y_2)

0.020793332819671106

In [10]:
from statsmodels.api import OLS
OLS(y_2,X_2).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                9Profit   R-squared (uncentered):                   0.188
Model:                            OLS   Adj. R-squared (uncentered):              0.188
Method:                 Least Squares   F-statistic:                              2447.
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:12   Log-Likelihood:                     -1.4889e+05
No. Observations:               21083   AIC:                                  2.978e+05
Df Residuals:                   21081   BIC:                                  2.978e+05
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
9Age          29.9746      0.459     65.258      0.000      29.074      30.875
9Online       35.2577      5.640      6.252      0.000      24.203      46.312
==============================================================================
Omnibus:                    10858.115   Durbin-Watson:                   1.986
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            76970.037
Skew:                           2.415   Prob(JB):                         0.00
Kurtosis:                      11.018   Cond. No.                         12.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### From the result from the OLS regression, the R-squared is 0.188,which is higher than 0.024.So age definitely help to explain if online channel has a significant impact on 1999 profitability.

# 3. To adjust for missing observations in the case of the variables 9Age and 9Inc (income), create the following variables::
- Substitute missing observations with zeros: Age0 and Inc0:
- Substitute missing observations with averages: AgeAvg and IncAvg:
- Include additional dummy variables where 1 if there is data and 0 otherwise: AgeExist and IncExis


In [11]:
data.head()

ID  9Profit  9Online  9Age  9Inc  9Tenure  9District  0Profit  0Online  \
0   1       21        0   NaN   NaN     6.33       1200      NaN      NaN   
1   2       -6        0   6.0   3.0    29.50       1200    -32.0      0.0   
2   3      -49        1   5.0   5.0    26.41       1100    -22.0      1.0   
3   4       -4        0   NaN   NaN     2.25       1200      NaN      NaN   
4   5      -61        0   2.0   9.0     9.91       1200     -4.0      0.0   

   9Billpay  0Billpay  
0         0       NaN  
1         0       0.0  
2         0       0.0  
3         0       NaN  
4         0       0.0

In [12]:
data['Age0'] = data['9Age'].fillna(0)
data['Inc0'] = data['9Inc'].fillna(0)
data['AgeAvg'] = data['9Age'].fillna(data['9Age'].mean())
data['IncAvg'] = data['9Inc'].fillna(data['9Inc'].mean())
data['AgeExist'] = (~pd.isnull(data['9Age'])).astype('int')
data['IncExis'] = (~pd.isnull(data['9Inc'])).astype('int')


In [13]:
data.head()

ID  9Profit  9Online  9Age  9Inc  9Tenure  9District  0Profit  0Online  \
0   1       21        0   NaN   NaN     6.33       1200      NaN      NaN   
1   2       -6        0   6.0   3.0    29.50       1200    -32.0      0.0   
2   3      -49        1   5.0   5.0    26.41       1100    -22.0      1.0   
3   4       -4        0   NaN   NaN     2.25       1200      NaN      NaN   
4   5      -61        0   2.0   9.0     9.91       1200     -4.0      0.0   

   9Billpay  0Billpay  Age0  Inc0    AgeAvg    IncAvg  AgeExist  IncExis  
0         0       NaN   0.0   0.0  4.046048  5.458777         0        0  
1         0       0.0   6.0   3.0  6.000000  3.000000         1        1  
2         0       0.0   5.0   5.0  5.000000  5.000000         1        1  
3         0       NaN   0.0   0.0  4.046048  5.458777         0        0  
4         0       0.0   2.0   9.0  2.000000  9.000000         1        1

To test for bias of missing data, evaluate if missing data has an effect on profitability analysis: 
- 3a. Evaluate the effect of online channel on 1999 profits when Age0 is included.
- 3b. Evaluate if adjusting missing data using Age0 or AgeAvg is relevant. In both cases, it is still necessary to include the additional variable AgeExist to control for the missing data
- 3c. Repeat above steps with income. Evaluate if adjusting missing data using Inc0 or IncAvg is relevant. Include AgeExist and AgeAvg in the calculations.

# 3 (a) Evaluate the effect of online channel on 1999 profits when Age0 is included.

In [14]:
from sklearn.linear_model import LinearRegression
X_3 = data[['Age0','9Online']]
y_3 = data['9Profit']
reg = LinearRegression().fit(X_3, y_3)
reg.score(X_3,y_3)

0.021611224887306046

In [15]:
from statsmodels.api import OLS
OLS(y_3,X_3).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                9Profit   R-squared (uncentered):                   0.149
Model:                            OLS   Adj. R-squared (uncentered):              0.149
Method:                 Least Squares   F-statistic:                              2767.
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:13   Log-Likelihood:                     -2.2221e+05
No. Observations:               31634   AIC:                                  4.444e+05
Df Residuals:                   31632   BIC:                                  4.444e+05
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Age0          29.1706      0.420     69.458      0.000      28.347      29.994
9Online       41.0307      4.513      9.091      0.000      32.184      49.877
==============================================================================
Omnibus:                    17831.541   Durbin-Watson:                   1.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           160396.392
Skew:                           2.611   Prob(JB):                         0.00
Kurtosis:                      12.717   Cond. No.                         11.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### When Age0 is included,the R-squared is 0.149,also much larger than 0.024. So Age0 is helpful to help explain if online channel has a significant impact on 1999 profitability. But it not so well as the base line

# 3 (b)Evaluate if adjusting missing data using Age0 or AgeAvg is relevant. In both cases, it is still necessary to include the additional variable AgeExist to control for the missing data

In [16]:
from sklearn.linear_model import LinearRegression
X_3_b = data[['AgeAvg','AgeExist','9Online']]
y_3_b = data['9Profit']
reg = LinearRegression().fit(X_3_b, y_3_b)
reg.score(X_3_b,y_3_b)

0.02425799852500099

In [17]:
from statsmodels.api import OLS
OLS(y_3_b,X_3_b).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                9Profit   R-squared (uncentered):                   0.163
Model:                            OLS   Adj. R-squared (uncentered):              0.163
Method:                 Least Squares   F-statistic:                              2052.
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:14   Log-Likelihood:                     -2.2195e+05
No. Observations:               31634   AIC:                                  4.439e+05
Df Residuals:                   31631   BIC:                                  4.439e+05
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
AgeAvg        20.1289      0.606     33.224      0.000      18.941      21.316
AgeExist      41.9879      3.053     13.752      0.000      36.004      47.972
9Online       13.6932      4.584      2.987      0.003       4.708      22.678
==============================================================================
Omnibus:                    18556.561   Durbin-Watson:                   1.996
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           174041.909
Skew:                           2.737   Prob(JB):                         0.00
Kurtosis:                      13.103   Cond. No.                         13.2
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### The R-squared is 0.163 which is larger than the base line.So it is relevant to use AgeAvg .

# 3 (c) Repeat above steps with income. Evaluate if adjusting missing data using Inc0 or IncAvg is relevant. Include AgeExist and AgeAvg in the calculations.

In [18]:
data.head()

ID  9Profit  9Online  9Age  9Inc  9Tenure  9District  0Profit  0Online  \
0   1       21        0   NaN   NaN     6.33       1200      NaN      NaN   
1   2       -6        0   6.0   3.0    29.50       1200    -32.0      0.0   
2   3      -49        1   5.0   5.0    26.41       1100    -22.0      1.0   
3   4       -4        0   NaN   NaN     2.25       1200      NaN      NaN   
4   5      -61        0   2.0   9.0     9.91       1200     -4.0      0.0   

   9Billpay  0Billpay  Age0  Inc0    AgeAvg    IncAvg  AgeExist  IncExis  
0         0       NaN   0.0   0.0  4.046048  5.458777         0        0  
1         0       0.0   6.0   3.0  6.000000  3.000000         1        1  
2         0       0.0   5.0   5.0  5.000000  5.000000         1        1  
3         0       NaN   0.0   0.0  4.046048  5.458777         0        0  
4         0       0.0   2.0   9.0  2.000000  9.000000         1        1

In [19]:
from sklearn.linear_model import LinearRegression
X_3_c = data[['Inc0','AgeExist','9Online','AgeAvg']]
y_3_c = data['9Profit']
reg = LinearRegression().fit(X_3_c, y_3_c)
reg.score(X_3_c,y_3_c)

0.0421613451066164

In [20]:
from statsmodels.api import OLS
OLS(y_3_c,X_3_c).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                9Profit   R-squared (uncentered):                   0.178
Model:                            OLS   Adj. R-squared (uncentered):              0.178
Method:                 Least Squares   F-statistic:                              1710.
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:15   Log-Likelihood:                     -2.2166e+05
No. Observations:               31634   AIC:                                  4.433e+05
Df Residuals:                   31630   BIC:                                  4.434e+05
Df Model:                           4                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Inc0          16.3703      0.684     23.929      0.000      15.029      17.711
AgeExist     -42.9367      4.664     -9.206      0.000     -52.078     -33.795
9Online        4.8413      4.558      1.062      0.288      -4.093      13.776
AgeAvg        19.7013      0.601     32.797      0.000      18.524      20.879
==============================================================================
Omnibus:                    18446.883   Durbin-Watson:                   1.997
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           173367.921
Skew:                           2.714   Prob(JB):                         0.00
Kurtosis:                      13.103   Cond. No.                         19.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [21]:
from sklearn.linear_model import LinearRegression
X_3_c_2 = data[['IncAvg','AgeExist','9Online','AgeAvg']]
y_3_c_2 = data['9Profit']
reg = LinearRegression().fit(X_3_c_2, y_3_c_2)
reg.score(X_3_c_2,y_3_c_2)

0.04298604654377214

In [22]:
from statsmodels.api import OLS
OLS(y_3_c_2,X_3_c_2).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                9Profit   R-squared (uncentered):                   0.169
Model:                            OLS   Adj. R-squared (uncentered):              0.169
Method:                 Least Squares   F-statistic:                              1611.
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:16   Log-Likelihood:                     -2.2183e+05
No. Observations:               31634   AIC:                                  4.437e+05
Df Residuals:                   31630   BIC:                                  4.437e+05
Df Model:                           4                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
IncAvg         9.1408      0.588     15.537      0.000       7.988      10.294
AgeExist      25.5827      3.220      7.946      0.000      19.272      31.893
9Online        0.7202      4.643      0.155      0.877      -8.379       9.820
AgeAvg        12.1977      0.790     15.431      0.000      10.648      13.747
==============================================================================
Omnibus:                    18701.637   Durbin-Watson:                   1.996
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           177202.331
Skew:                           2.762   Prob(JB):                         0.00
Kurtosis:                      13.195   Cond. No.                         21.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### The result of the R-squared show that adjusting missing data using Inc0 or IncAvg is relevant. Because they both increase the R-squared.

# 4. Evaluate if online channel has a significant impact on 1999 profitability after controlling for demographic variables: age, income, tenure, and geographic district.
To evaluate the latter, create dummy variables D1100 and D1200 for districts 1100 and 1200 respectively from the variable 9District.

In [23]:
data.head()

ID  9Profit  9Online  9Age  9Inc  9Tenure  9District  0Profit  0Online  \
0   1       21        0   NaN   NaN     6.33       1200      NaN      NaN   
1   2       -6        0   6.0   3.0    29.50       1200    -32.0      0.0   
2   3      -49        1   5.0   5.0    26.41       1100    -22.0      1.0   
3   4       -4        0   NaN   NaN     2.25       1200      NaN      NaN   
4   5      -61        0   2.0   9.0     9.91       1200     -4.0      0.0   

   9Billpay  0Billpay  Age0  Inc0    AgeAvg    IncAvg  AgeExist  IncExis  
0         0       NaN   0.0   0.0  4.046048  5.458777         0        0  
1         0       0.0   6.0   3.0  6.000000  3.000000         1        1  
2         0       0.0   5.0   5.0  5.000000  5.000000         1        1  
3         0       NaN   0.0   0.0  4.046048  5.458777         0        0  
4         0       0.0   2.0   9.0  2.000000  9.000000         1        1

In [24]:
data_dropna = data.dropna(axis = 0,how = 'any')


In [25]:
data_dropna.loc[:,'9District'] = data_dropna.loc[:,'9District'].astype('str')
data_dum = pd.get_dummies(data_dropna)
data_dum.head()

/Users/lilinsen/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


ID  9Profit  9Online  9Age  9Inc  9Tenure  0Profit  0Online  9Billpay  \
1   2       -6        0   6.0   3.0    29.50    -32.0      0.0         0   
2   3      -49        1   5.0   5.0    26.41    -22.0      1.0         0   
4   5      -61        0   2.0   9.0     9.91     -4.0      0.0         0   
6   7      -19        0   3.0   1.0     8.41      0.0      0.0         0   
7   8       59        0   5.0   8.0     7.33    -65.0      0.0         0   

   0Billpay  Age0  Inc0  AgeAvg  IncAvg  AgeExist  IncExis  9District_1100  \
1       0.0   6.0   3.0     6.0     3.0         1        1               0   
2       0.0   5.0   5.0     5.0     5.0         1        1               1   
4       0.0   2.0   9.0     2.0     9.0         1        1               0   
6       0.0   3.0   1.0     3.0     1.0         1        1               0   
7       0.0   5.0   8.0     5.0     8.0         1        1               0   

   9District_1200  9District_1300  
1               1               0  
2               0               0  
4               1               0  
6               0               1  
7               1               0

In [26]:

from sklearn.linear_model import LinearRegression
X_4 = data_dum[['9Age','9Inc','9Tenure','9District_1100','9District_1200','9Online']]
y_4 = data_dum['9Profit']
reg = LinearRegression().fit(X_4, y_4)
reg.score(X_4,y_4)

0.05869494513034579

In [27]:
from statsmodels.api import OLS
OLS(y_4,X_4).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                9Profit   R-squared (uncentered):                   0.215
Model:                            OLS   Adj. R-squared (uncentered):              0.215
Method:                 Least Squares   F-statistic:                              961.5
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:17   Log-Likelihood:                     -1.4854e+05
No. Observations:               21083   AIC:                                  2.971e+05
Df Residuals:                   21077   BIC:                                  2.971e+05
Df Model:                           6                                                  
Covariance Type:            nonrobust                                                  
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
9Age               9.8455      1.099      8.958      0.000       7.691      12.000
9Inc              12.6969      0.734     17.303      0.000      11.259      14.135
9Tenure            3.9646      0.246     16.092      0.000       3.482       4.448
9District_1100   -46.6832      7.483     -6.238      0.000     -61.351     -32.015
9District_1200   -21.3579      5.009     -4.264      0.000     -31.176     -11.539
9Online           10.1770      5.765      1.765      0.078      -1.123      21.477
==============================================================================
Omnibus:                    10949.099   Durbin-Watson:                   1.986
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            79768.420
Skew:                           2.429   Prob(JB):                         0.00
Kurtosis:                      11.198   Cond. No.                         64.6
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### The p_value of online channel is 0.078, which is higher than other variable.So the online channel doesn't have a significant impact on 1999 profitability after controlling for demographic variables: age, income, tenure, and geographic district.¶

# 5 (a) Evaluate the drivers of customer profitability for the year 2000 (Hint: you can evaluate the variables explored for profitability of 1999).

In [28]:
data_dropna = data.dropna(axis = 0,how = 'any')
data_dropna.head()

ID  9Profit  9Online  9Age  9Inc  9Tenure  9District  0Profit  0Online  \
1   2       -6        0   6.0   3.0    29.50       1200    -32.0      0.0   
2   3      -49        1   5.0   5.0    26.41       1100    -22.0      1.0   
4   5      -61        0   2.0   9.0     9.91       1200     -4.0      0.0   
6   7      -19        0   3.0   1.0     8.41       1300      0.0      0.0   
7   8       59        0   5.0   8.0     7.33       1200    -65.0      0.0   

   9Billpay  0Billpay  Age0  Inc0  AgeAvg  IncAvg  AgeExist  IncExis  
1         0       0.0   6.0   3.0     6.0     3.0         1        1  
2         0       0.0   5.0   5.0     5.0     5.0         1        1  
4         0       0.0   2.0   9.0     2.0     9.0         1        1  
6         0       0.0   3.0   1.0     3.0     1.0         1        1  
7         0       0.0   5.0   8.0     5.0     8.0         1        1

In [29]:
from sklearn.linear_model import LinearRegression
X_5_a = data_dum[['9Age','9Inc','9Tenure','9District_1100','9District_1200','9Online']]
y_5_a = data_dum['0Profit']
reg = LinearRegression().fit(X_5_a, y_5_a)
reg.score(X_5_a,y_5_a)

0.03354586918818114

In [30]:
from statsmodels.api import OLS
OLS(y_5_a,X_5_a).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                0Profit   R-squared (uncentered):                   0.159
Model:                            OLS   Adj. R-squared (uncentered):              0.159
Method:                 Least Squares   F-statistic:                              663.8
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:21   Log-Likelihood:                     -1.5570e+05
No. Observations:               21083   AIC:                                  3.114e+05
Df Residuals:                   21077   BIC:                                  3.115e+05
Df Model:                           6                                                  
Covariance Type:            nonrobust                                                  
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
9Age               7.6215      1.544      4.937      0.000       4.595      10.648
9Inc              17.8734      1.031     17.340      0.000      15.853      19.894
9Tenure            3.9607      0.346     11.445      0.000       3.282       4.639
9District_1100   -44.8677     10.511     -4.268      0.000     -65.471     -24.265
9District_1200   -17.1606      7.036     -2.439      0.015     -30.952      -3.369
9Online           15.7686      8.098      1.947      0.052      -0.104      31.642
==============================================================================
Omnibus:                    45586.221   Durbin-Watson:                   1.979
Prob(Omnibus):                  0.000   Jarque-Bera (JB):       1065165629.965
Skew:                          18.703   Prob(JB):                         0.00
Kurtosis:                    1103.518   Cond. No.                         64.6
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### From the summery of the ols regression. All 1999 variable except 9Online are significant.

# 5 (b) Evaluate if the variable 9Profit should be included in the customer profitability analysis for 2000.

In [31]:
from sklearn.linear_model import LinearRegression
X_5_b = data_dum[['9Age','9Inc','9Tenure','9District_1100','9District_1200','9Online','9Profit']]
y_5_b = data_dum['0Profit']
reg = LinearRegression().fit(X_5_b, y_5_b)
reg.score(X_5_b,y_5_b)

0.3728636372189129

In [32]:
from statsmodels.api import OLS
OLS(y_5_b,X_5_b).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                0Profit   R-squared (uncentered):                   0.456
Model:                            OLS   Adj. R-squared (uncentered):              0.455
Method:                 Least Squares   F-statistic:                              2519.
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:22   Log-Likelihood:                     -1.5112e+05
No. Observations:               21083   AIC:                                  3.023e+05
Df Residuals:                   21076   BIC:                                  3.023e+05
Df Model:                           7                                                  
Covariance Type:            nonrobust                                                  
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
9Age              -0.5908      1.245     -0.475      0.635      -3.030       1.849
9Inc               7.2826      0.835      8.720      0.000       5.646       8.920
9Tenure            0.6537      0.280      2.333      0.020       0.105       1.203
9District_1100    -5.9283      8.465     -0.700      0.484     -22.521      10.665
9District_1200     0.6544      5.664      0.116      0.908     -10.447      11.756
9Online            7.2798      6.516      1.117      0.264      -5.493      20.052
9Profit            0.8341      0.008    107.146      0.000       0.819       0.849
==============================================================================
Omnibus:                    57209.668   Durbin-Watson:                   1.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):       6079635564.109
Skew:                          32.990   Prob(JB):                         0.00
Kurtosis:                    2632.914   Cond. No.                     1.32e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.32e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### From the result, we see the t statistics for 9Profit is extremely high.At the same time, the R-squared raises significantly.That is because we include the variable 9Profit.So the variable 9Profit should be included in the customer profitability analysis for 2000.

# 5 (c) Evaluate the drivers of customer profitability for 2000 after adding electronic billpay using OLS and Random Forest regression. Compare these two methods based on their R-squared.

In [33]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
X_5_c = data_dum[['9Age','9Inc','9Tenure','9District_1100','9District_1200','9Online','9Billpay','9Profit']]
y_5_c = data_dum['0Profit']
reg = LinearRegression().fit(X_5_c, y_5_c)
rfr = RandomForestRegressor(max_depth=2, random_state=0,n_estimators=100).fit(X_5_c, y_5_c) 
reg.score(X_5_c,y_5_c)

0.37287147567985834

In [34]:
rfr.score(X_5_c,y_5_c)

0.35900444736034065

In [35]:
from statsmodels.api import OLS
OLS(y_5_c,X_5_c).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                0Profit   R-squared (uncentered):                   0.456
Model:                            OLS   Adj. R-squared (uncentered):              0.455
Method:                 Least Squares   F-statistic:                              2204.
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:24   Log-Likelihood:                     -1.5112e+05
No. Observations:               21083   AIC:                                  3.023e+05
Df Residuals:                   21075   BIC:                                  3.023e+05
Df Model:                           8                                                  
Covariance Type:            nonrobust                                                  
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
9Age              -0.5867      1.245     -0.471      0.637      -3.026       1.853
9Inc               7.2790      0.835      8.715      0.000       5.642       8.916
9Tenure            0.6532      0.280      2.332      0.020       0.104       1.202
9District_1100    -5.8737      8.466     -0.694      0.488     -22.468      10.721
9District_1200     0.6805      5.664      0.120      0.904     -10.422      11.783
9Online            6.0573      6.943      0.872      0.383      -7.552      19.666
9Billpay           8.9185     17.481      0.510      0.610     -25.345      43.182
9Profit            0.8339      0.008    107.023      0.000       0.819       0.849
==============================================================================
Omnibus:                    57210.256   Durbin-Watson:                   1.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):       6079953553.996
Skew:                          32.991   Prob(JB):                         0.00
Kurtosis:                    2632.983   Cond. No.                     2.56e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.56e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### The result is that the R-squared of OLS is larger than Random Forest regression's. The p_value also show that when we have 9Profit as the feature, other variable are not significant anymore.

# 6 (a) Evaluate the drivers of customer retention for the year 2000. Compare OLS with logistic regression results and indicate which one of these two methods is more appropriate for this problem. Retain takes a value of 0 when 0Profit has a missing observation and 1 otherwise.

In [36]:
data['Retain'] = (~data['0Profit'].isnull()).astype('int')

In [37]:
data.loc[:,'9District'] = data_dropna.loc[:,'9District'].astype('str')
data = pd.get_dummies(data)


In [38]:
data.head()

ID  9Profit  9Online  9Age  9Inc  9Tenure  0Profit  0Online  9Billpay  \
0   1       21        0   NaN   NaN     6.33      NaN      NaN         0   
1   2       -6        0   6.0   3.0    29.50    -32.0      0.0         0   
2   3      -49        1   5.0   5.0    26.41    -22.0      1.0         0   
3   4       -4        0   NaN   NaN     2.25      NaN      NaN         0   
4   5      -61        0   2.0   9.0     9.91     -4.0      0.0         0   

   0Billpay  Age0  Inc0    AgeAvg    IncAvg  AgeExist  IncExis  Retain  \
0       NaN   0.0   0.0  4.046048  5.458777         0        0       0   
1       0.0   6.0   3.0  6.000000  3.000000         1        1       1   
2       0.0   5.0   5.0  5.000000  5.000000         1        1       1   
3       NaN   0.0   0.0  4.046048  5.458777         0        0       0   
4       0.0   2.0   9.0  2.000000  9.000000         1        1       1   

   9District_1100  9District_1200  9District_1300  
0               0               0               0  
1               0               1               0  
2               1               0               0  
3               0               0               0  
4               0               1               0

In [39]:
X_6_a = data[['AgeAvg','IncAvg','9Tenure','9District_1100','9District_1200','9Online','9Billpay','9Profit']]
y_6_a = data['Retain']

In [40]:
from statsmodels.api import OLS
OLS(y_6_a,X_6_a).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                 Retain   R-squared (uncentered):                   0.858
Model:                            OLS   Adj. R-squared (uncentered):              0.858
Method:                 Least Squares   F-statistic:                          2.386e+04
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:12:25   Log-Likelihood:                         -11166.
No. Observations:               31634   AIC:                                  2.235e+04
Df Residuals:                   31626   BIC:                                  2.242e+04
Df Model:                           8                                                  
Covariance Type:            nonrobust                                                  
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
AgeAvg             0.0699      0.001     62.855      0.000       0.068       0.072
IncAvg             0.0417      0.001     55.156      0.000       0.040       0.043
9Tenure            0.0032      0.000     12.753      0.000       0.003       0.004
9District_1100     0.4745      0.008     57.458      0.000       0.458       0.491
9District_1200     0.4367      0.004    107.717      0.000       0.429       0.445
9Online            0.0743      0.006     11.729      0.000       0.062       0.087
9Billpay          -0.0718      0.016     -4.445      0.000      -0.103      -0.040
9Profit        -3.769e-05    7.3e-06     -5.166      0.000    -5.2e-05   -2.34e-05
==============================================================================
Omnibus:                     1627.036   Durbin-Watson:                   1.973
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              771.080
Skew:                          -0.184   Prob(JB):                    3.65e-168
Kurtosis:                       2.329   Cond. No.                     2.48e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.48e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [41]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(random_state=0, solver='lbfgs',max_iter = 1000,multi_class='multinomial').fit(X_6_a, y_6_a)
clf.score(X_6_a, y_6_a)

0.8349876714926977

### In this case, the R-squared of OLS regression is 0.858, which is very high. At the same time, the accuracy for the LogisticRegression is 83.5%. Since this is a binary classification, so the LogisticRegression is more suitable in this case.

# 6 (b) Calculate and rank the odds ratio (exp(coefficients of logistic regression)) of the logistic regression and explain the impact of the 9Online and Billpay variables on customer's retention.

In [42]:
import numpy as np
import statsmodels.formula.api as sm
X_6_a = data[['AgeAvg','IncAvg','9Tenure','9District_1100','9District_1200','9Online','9Billpay','9Profit']]
y_6_a = data['Retain']
fea = ['AgeAvg','IncAvg','9Tenure','9District_1100','9District_1200','9Online','9Billpay','9Profit']

In [43]:
np.exp(clf.coef_)

array([[ 1.01783954,  0.9643891 ,  1.01842294, 15.20576293, 53.94829618,
         1.08973172,  0.65544826,  1.0000669 ]])

In [44]:
s = pd.DataFrame()
s['variable'] = fea
s['odds ratio'] = [ 1.01783954,  0.9643891 ,  1.01842294, 15.20576293, 53.94829618,
         1.08973172,  0.65544826,  1.0000669 ]
s.sort_values(by=['odds ratio'])

variable  odds ratio
6        9Billpay    0.655448
1          IncAvg    0.964389
7         9Profit    1.000067
0          AgeAvg    1.017840
2         9Tenure    1.018423
5         9Online    1.089732
3  9District_1100   15.205763
4  9District_1200   53.948296

### - The odds ratio of 9Online is 1.09, means if the 9Online is 1 instead of 0, there will 1.09 times greater odds of being retention. 
### - The odds ratio of 9Billpay is 0.65, means if the 9Billpay is 1 instead of 0, there will 0.65 times less odds of being retention. 

# 7. Evaluate the effect of the online channel and billpay on customer's retention using a hidden Markov model (HMM) with the variables 9Online, 9Billpay, 0Online, 0Billpay and the new variable Retain. Retain takes a value of 0 when 0Profit has a missing observation and 1 otherwise. (You must install the package hmmlearn. This link includes the installation instructions and documentation: https://github.com/hmmlearn/hmmlearn )

In [45]:
data

ID  9Profit  9Online  9Age  9Inc  9Tenure  0Profit  0Online  \
0          1       21        0   NaN   NaN     6.33      NaN      NaN   
1          2       -6        0   6.0   3.0    29.50    -32.0      0.0   
2          3      -49        1   5.0   5.0    26.41    -22.0      1.0   
3          4       -4        0   NaN   NaN     2.25      NaN      NaN   
4          5      -61        0   2.0   9.0     9.91     -4.0      0.0   
...      ...      ...      ...   ...   ...      ...      ...      ...   
31629  31630      -50        0   5.0   5.0     3.75      1.0      0.0   
31630  31631      458        0   3.0   8.0    12.08    423.0      1.0   
31631  31632      -83        0   6.0   4.0    15.83    -60.0      0.0   
31632  31633       92        1   1.0   6.0     5.41    170.0      1.0   
31633  31634      124        0   3.0   6.0    17.50    150.0      0.0   

       9Billpay  0Billpay  Age0  Inc0    AgeAvg    IncAvg  AgeExist  IncExis  \
0             0       NaN   0.0   0.0  4.046048  5.458777         0        0   
1             0       0.0   6.0   3.0  6.000000  3.000000         1        1   
2             0       0.0   5.0   5.0  5.000000  5.000000         1        1   
3             0       NaN   0.0   0.0  4.046048  5.458777         0        0   
4             0       0.0   2.0   9.0  2.000000  9.000000         1        1   
...         ...       ...   ...   ...       ...       ...       ...      ...   
31629         0       0.0   5.0   5.0  5.000000  5.000000         1        1   
31630         0       0.0   3.0   8.0  3.000000  8.000000         1        1   
31631         0       0.0   6.0   4.0  6.000000  4.000000         1        1   
31632         0       0.0   1.0   6.0  1.000000  6.000000         1        1   
31633         0       0.0   3.0   6.0  3.000000  6.000000         1        1   

       Retain  9District_1100  9District_1200  9District_1300  
0           0               0               0               0  
1           1               0               1               0  
2           1               1               0               0  
3           0               0               0               0  
4           1               0               1               0  
...       ...             ...             ...             ...  
31629       1               0               1               0  
31630       1               0               0               1  
31631       1               0               1               0  
31632       1               0               1               0  
31633       1               0               0               1  

[31634 rows x 20 columns]

In [46]:
import pandas as pd
import numpy as np

from hmmlearn.hmm import GaussianHMM

Online_9 = data['9Online']
Billpay_9 = data['9Billpay']
Online_0 = data['0Online'].fillna(0)
Billpay_0 = data['0Billpay'].fillna(0)
Retain = data['Retain']
X_7 = np.column_stack([Online_9, Billpay_9,Billpay_0,Online_0,Retain])


In [47]:
# Run Gaussian HMM
print("fitting to HMM and decoding ...", end='')
n_components = 2
# make an HMM instance and execute fit
model = GaussianHMM(n_components, covariance_type="full", n_iter=100)
model.fit(X_7)
# predict the optimal sequence of internal hidden state
# hidden_states = model.predict(X_7)

print("done\n")

fitting to HMM and decoding ...done



In [48]:
print("Transition matrix")
print(model.transmat_)
print()

print("means and vars of each hidden state")
for i in range(n_components):
    print("%dth hidden state" % i)
    print("mean = ", model.means_[i])
    print("var = ", np.diag(model.covars_[i]))
    print()

Transition matrix
[[0.82952642 0.17047358]
 [0.83466915 0.16533085]]

means and vars of each hidden state
0th hidden state
mean =  [0.02908371 0.         0.         0.         0.80437017]
var =  [2.82382293e-02 3.80676843e-07 3.80676843e-07 3.80676843e-07
 1.57359180e-01]

1th hidden state
mean =  [0.57595527 0.09841566 0.14874185 0.97968313 0.98154706]
var =  [0.24423266 0.08873188 0.12661957 0.01990596 0.01811429]



# 8. Build a transition matrix (online, billpay) from 1999 to 2000 from those different customers' states:those that were online, offline without electronic billpay, online with electronic billpay for 2000, customers who left the bank. Explain the billpay effect on customers' retention.



In [69]:
data.head()

ID  9Profit  9Online  9Age  9Inc  9Tenure  0Profit  0Online  9Billpay  \
0   1       21        0   NaN   NaN     6.33      NaN      NaN         0   
1   2       -6        0   6.0   3.0    29.50    -32.0      0.0         0   
2   3      -49        1   5.0   5.0    26.41    -22.0      1.0         0   
3   4       -4        0   NaN   NaN     2.25      NaN      NaN         0   
4   5      -61        0   2.0   9.0     9.91     -4.0      0.0         0   

   0Billpay  Age0  Inc0    AgeAvg    IncAvg  AgeExist  IncExis  Retain  \
0       NaN   0.0   0.0  4.046048  5.458777         0        0       0   
1       0.0   6.0   3.0  6.000000  3.000000         1        1       1   
2       0.0   5.0   5.0  5.000000  5.000000         1        1       1   
3       NaN   0.0   0.0  4.046048  5.458777         0        0       0   
4       0.0   2.0   9.0  2.000000  9.000000         1        1       1   

   9District_1100  9District_1200  9District_1300  
0               0               0               0  
1               0               1               0  
2               1               0               0  
3               0               0               0  
4               0               1               0

In [53]:
data['online_false_elc_false'] = 1 - (data['0Online']*data['0Billpay'])
data['online_true_elc_false'] = data['0Online']*(1-data['0Billpay'])
data['online_true_elc_true'] = data['0Online']*data['0Billpay']
data['left'] = 1 - data['Retain']
X_8_1 = data['online_false_elc_false']
X_8_2 = data['online_true_elc_false']
X_8_3 = data['online_true_elc_true']
X_8_4 = data['left'] = 1 - data['Retain']
df = data[['online_false_elc_false','online_true_elc_false','online_true_elc_true','left']].fillna(0)
X_8 = np.column_stack([X_8_1, X_8_2,X_8_3,X_8_4])
# Run Gaussian HMM
print("fitting to HMM and decoding ...", end='')
n_components = 4
# make an HMM instance and execute fit
model = GaussianHMM(n_components, covariance_type="full", n_iter=100)
model.fit(df)
# predict the optimal sequence of internal hidden state
# hidden_states = model.predict(X_7)

print("done\n")
print("Transition matrix")
print(model.transmat_)
print()

fitting to HMM and decoding ...done

Transition matrix
[[0.67060547 0.14132438 0.16415371 0.02391643]
 [0.67436127 0.14096818 0.16024204 0.02442851]
 [0.65453152 0.14293926 0.17244683 0.03008239]
 [0.68553459 0.12201258 0.16352201 0.02893082]]



In [54]:
Transition_matrix = model.transmat_
Transition_matrix.dot(Transition_matrix).dot(Transition_matrix).dot(Transition_matrix).dot(Transition_matrix)

array([[0.66885898, 0.14105518, 0.16495406, 0.02513178],
       [0.66885898, 0.14105518, 0.16495406, 0.02513178],
       [0.66885898, 0.14105518, 0.16495406, 0.02513179],
       [0.66885898, 0.14105518, 0.16495406, 0.02513178]])

### We time the Transition_matrix for 4 times and find that it comes to stable. The the last probability of these four state is 0.67,0.14,0.16,0.03, each represent those that were online, offline without electronic billpay, online with electronic billpay for 2000, customers who left the bank. 